src/mumble/AudioPreprocessor.h

```diff
@@ -0,0 +1,215 @@
+// Copyright 2024 The Mumble Developers. All rights reserved.
+// Use of this source code is governed by a BSD-style license
+// that can be found in the LICENSE file at the root of the
+// Mumble source tree or at <https://www.mumble.info/LICENSE>.
+
+#ifndef MUMBLE_MUMBLE_AUDIOPREPROCESSOR_H_
+#define MUMBLE_MUMBLE_AUDIOPREPROCESSOR_H_
+
+#include <cstdint>
+#include <vector>
+
+struct SpeexEchoState_;
+struct SpeexPreprocessState_;
+
+class AudioPreprocessor {
+public:
+	using psd_t = std::vector< std::int32_t >;
+
+	/**
+	 * \return Whether the object is initialized and all its methods can be safely called.
+	 */
+	constexpr explicit operator bool() const { return m_handle != nullptr; }
+
+	AudioPreprocessor() = default;
+	AudioPreprocessor(AudioPreprocessor &&other);
+	~AudioPreprocessor();
+
+	AudioPreprocessor &operator=(AudioPreprocessor &&other);
+
+	/**
+	 * Initializes the object. This must be called at least once before any other function!
+	 *
+	 * \param sampleRate The sample rate in Hz.
+	 *
+	 * \param quantum The number of samples that are expected for each preprocess cycle.
+	 *                Should correspond to 10-20 ms (480-960 with a sample rate of 48000).
+	 *
+	 * \return Whether initialization succeeded.
+	 */
+	bool init(std::uint32_t sampleRate, std::uint32_t quantum);
+	/**
+	 * Deinitializes the object, doesn't do anything if it's not initialized.
+	 * Once that is done no function other than init() can be called.
+	 */
+	void deinit();
+
+	/**
+	 * Runs the preprocessor, which analyzes and modifies the samples.
+	 *
+	 * \param samples The samples that should be processed (modified in-place).
+	 *
+	 * \return Whether speech was detected, if VAD is enabled. Always 'true' otherwise.
+	 */
+	bool run(std::int16_t &samples);
+
+	/**
+	 * \return Handle to the echo state object.
+	 */
+	SpeexEchoState_ *getEchoState();
+	/**
+	 * Set the echo state object, to toggle the echo canceller.
+	 *
+	 * \param handle Handle to the echo state. Passing 'nullptr' disables the feature.
+	 *
+	 * \return Whether the operation succeeded.
+	 */
+	bool setEchoState(SpeexEchoState_ *handle);
+
+	/**
+	 * \return Whether automatic gain control is enabled.
+	 */
+	bool usesAGC() const;
+	/**
+	 * Toggles automatic gain control.
+	 *
+	 * \param enable 'true' to enable the feature, otherwise 'false'.
+	 *
+	 * \return Whether the operation succeeded.
+	 */
+	bool toggleAGC(bool enable);
```

Comment:
All these functions don't _toggle_, they _set_ the value. Imo they should be renamed appropriately.